server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java

```diff
@@ -0,0 +1,257 @@
+                clusterService.state().nodes().getLocalNodeId()
+            )
+        );
+    }
+
+    /**
+     * The cluster state indicating that a node is to be shutdown may be cleared instead of following through with node shutdown. In that
+     * case, no further shutdown shard snapshot progress reporting is desired.
+     */
+    public void onClusterStateRemoveShutdown() {
+        assert shutdownStartMillis != -1 : "Expected a call to add shutdown mode before a call to remove shutdown mode.";
+
+        // Reset the shutdown specific trackers.
+        this.shutdownStartMillis = -1;
+        this.shutdownFinishedSignallingPausingMillis = -1;
+
+        // Turn off the progress logger, which we only want to run during shutdown.
+        cancelProgressLogger();
+    }
+
+    /**
+     * Tracks how many shard snapshots are started.
+     */
+    public void incNumberOfShardSnapshotsInProgress() {
+        logger.trace("New shard snapshot started");
+        numberOfShardSnapshotsInProgressOnDataNode.incrementAndGet();
+    }
+
+    /**
+     * Tracks how many shard snapshots have finished since shutdown mode began.
+     */
+    public void decNumberOfShardSnapshotsInProgress(IndexShardSnapshotStatus.Stage stage) {
+        logger.trace(Strings.format("A shard snapshot finished in state [%s]", stage));
+        numberOfShardSnapshotsInProgressOnDataNode.decrementAndGet();
+        if (shutdownStartMillis != -1) {
+            switch (stage) {
+                case DONE -> doneCount.incrementAndGet();
+                case FAILURE -> failureCount.incrementAndGet();
+                case ABORTED -> abortedCount.incrementAndGet();
+                case PAUSED -> pausedCount.incrementAndGet();
+                // The other stages are active, we should only see the end result because this method is called upon completion.
+                default -> {
+                    assert false : "unexpected shard snapshot stage: " + stage;
+                }
+            }
+        }
+    }
+
+    /**
+     * Uniquely tracks a request to update a shard snapshot status sent to the master node. Idempotent, safe to call multiple times.
+     *
+     * @param snapshot first part of a unique identifier
+     * @param shardId second part of a unique identifier
+     */
+    public void trackRequestSentToMaster(Snapshot snapshot, ShardId shardId) {
+        logger.trace("Tracking shard snapshot request to master");
+        shardSnapshotRequests.add(snapshot.toString() + shardId.getIndexName() + shardId.getId());
+    }
+
+    /**
+     * Stops tracking a request to update a shard snapshot status sent to the master node. Idempotent, safe to call multiple times.
+     *
+     * @param snapshot first part of a unique identifier
+     * @param shardId second part of a unique identifier
+     */
+    public void releaseRequestSentToMaster(Snapshot snapshot, ShardId shardId) {
+        logger.trace("A shard snapshot request to master finished");
```

Comment:
I reckon we should report the snapshot and shard in this message. It might also be worth reporting the elapsed time since we sent the message?